### PR TITLE
Make it possible to read PropertyLists as persistables

### PR DESCRIPTION
**Pull request by RobertLuptonTheGood**

### CODE REVIEW

File: src/formatters/PropertyListFormatter.cc [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+{
+    LOGL_DEBUG(_log, "PropertyListFormatter write start");
+    auto ip = dynamic_cast<lsst::daf::base::PropertyList const*>(persistable);
+    if (ip == 0) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "Persisting non-PropertyList");
+    }
+    if (typeid(*storage) == typeid(lsst::daf::persistence::FitsStorage)) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError,
+                      "Writing a PropertyList to FitsStorage is not supported");
+    }
```
Comment (TallJimbo) [inline src/formatters/PropertyListFormatter.cc]: I was confused by this until I realized that *writing* (as opposed to *FitsStorage*) is what's not supported; might be clearer if you just say that `PropertyList` persistence is read-only.

File: src/formatters/PropertyListFormatter.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+PropertyListFormatter::~PropertyListFormatter(void) {
+}
+
+void PropertyListFormatter::write(
+        lsst::daf::base::Persistable const* persistable,
+        lsst::daf::persistence::Storage::Ptr storage,
+        lsst::daf::base::PropertySet::Ptr
+                                                )
```
Comment (TallJimbo) [inline src/formatters/PropertyListFormatter.cc]: More weird indentation and `::Ptr` usage here (and in most other signatures in this file).

File: src/formatters/PropertyListFormatter.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        fitsfile.setHdu(hdu);
+        
+        auto metadata = new lsst::daf::base::PropertyList;
+        fitsfile.readMetadata(*metadata, strip);
+        // if INHERIT=T, we want to also include header entries from the primary HDU
+        if (fitsfile.getHdu() != 0 && metadata->exists("INHERIT")) {
+            LOGLS_WARN(_log, "Ignoring \"INHERIT\" keyword in " << fileName);
+        }
+        return metadata;
+    }
```
Comment (TallJimbo) [inline src/formatters/PropertyListFormatter.cc]: I think it'd be better to have this function return by `std::unique_ptr<daf::base::PropertyList>` and then use `release()` at the end of `PropertyListFormatter::read`.

File: tests/testPropertyListPersistence.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        logicalLocation = dafPers.LogicalLocation(os.path.join("tests", "data", "HSC-0908120-056-small.fits"))
+
+        # Create a FitsStorage and put it in a StorageList.
+        storage = self.persistence.getRetrieveStorage("FitsStorage", logicalLocation)
+        storageList = dafPers.StorageList([storage])
+
+        # Let's do the retrieval!
+        propertyList = self.persistence.unsafeRetrieve("PropertyList", storageList, None)
+
+        assert propertyList.get("AR_HDU") == 5
```
Comment (TallJimbo) [inline tests/testPropertyListPersistence.py]: Should be `self.assertEqual(propertyList.get("AR_HDU"), 5)`.

File: include/lsst/afw/formatters/PropertyListFormatter.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        lsst::daf::base::PropertySet::Ptr additionalData
+    );
+    virtual void update(
+        lsst::daf::base::Persistable* persistable,
+        lsst::daf::persistence::Storage::Ptr storage,
+        lsst::daf::base::PropertySet::Ptr additionalData
+    );
+
+    static lsst::daf::persistence::Formatter::Ptr createInstance(
+        lsst::pex::policy::Policy::Ptr policy
```
Comment (RobertLuptonTheGood) [inline include/lsst/afw/formatters/PropertyListFormatter.h]: I changed all these to PTR() and had a template match error (that was actually caused by changing `PropertySet` to `PropertyList`).  I'll go through and change them all to `std::shared_ptr<T>`.

File: src/formatters/PropertyListFormatter.cc [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+ * @brief Implementation of PropertyListFormatter class
+ *
+ * @ingroup afw
+ */
+
+#ifndef __GNUC__
+#  define __attribute__(x) /*NOTHING*/
+#endif
+static char const* SVNid __attribute__((unused)) =
+    "$Id$";
```
Comment (RobertLuptonTheGood) [inline src/formatters/PropertyListFormatter.cc]: Yes; I didn't remove it as it was an unnecessary change to the` DecoratedImageFormatter`, but that hardly matters in a copy.

File: src/formatters/PropertyListFormatter.cc [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+namespace lsst {
+namespace afw {
+namespace formatters {
+
+lsst::daf::persistence::FormatterRegistration PropertyListFormatter::registration(
+    "PropertyList", typeid(lsst::daf::base::PropertyList), createInstance);
+
+PropertyListFormatter::PropertyListFormatter(
+        lsst::pex::policy::Policy::Ptr
+                                                             )
```
Comment (RobertLuptonTheGood) [inline src/formatters/PropertyListFormatter.cc]: 8 characters is copied from the `DecoratedImageFormatter`; what would you prefer?  The ) is weird and I've fixed it.

File: src/formatters/PropertyListFormatter.cc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        lsst::daf::base::Persistable const* persistable,
+        lsst::daf::persistence::Storage::Ptr storage,
+        lsst::daf::base::PropertySet::Ptr
+                                                )
+{
+    LOGL_DEBUG(_log, "PropertyListFormatter write start");
+    auto ip = dynamic_cast<lsst::daf::base::PropertyList const*>(persistable);
+    if (ip == 0) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "Persisting non-PropertyList");
+    }
```
Comment (RobertLuptonTheGood) [inline src/formatters/PropertyListFormatter.cc]: Cargo culted from `DecoratedImageFormatter`.  I'll leave it unless you insist.

File: src/formatters/PropertyListFormatter.cc (hunk 7 above)
Comment (TallJimbo) [inline src/formatters/PropertyListFormatter.cc]: 4 chars is standard.
